p2p/pex/addrbook_test.go

```diff
@@ -36,6 +37,264 @@ func deleteTempFile(fname string) {
 	}
 }
 
+func TestAddrBookGetSelectionWithOneMarkedGood(t *testing.T) {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	N := 10
+	randAddrs := randNetAddressPairs(t, N)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	// mark one addr as good
+	for _, addr := range randAddrs[:1] {
+		book.MarkGood(addr.addr)
+	}
+
+	// ensure we can get selection.
+	// this used to result in an infinite loop (!)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.NotNil(t, addrs, "expected an address")
+}
+
+func TestAddrBookGetSelectionWithOneNotMarkedGood(t *testing.T) {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	// 0 addresses
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	N := 10
+	randAddrs := randNetAddressPairs(t, N)
+	// mark all addrs but one as good
+	for _, addr := range randAddrs[:N-1] {
+		book.AddAddress(addr.addr, addr.src)
+		book.MarkGood(addr.addr)
+	}
+	// add the last addr, don't mark as good
+	for _, addr := range randAddrs[N-1:] {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	// ensure we can get selection.
+	// this used to result in an infinite loop (!)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.NotNil(t, addrs, "expected an address")
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func listsAreEqual(a, b []int) bool {
+	if (a == nil) != (b == nil) {
+		return false
+	}
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// Creates a Book with m old and n new addresses.
+func createAddrBook(t *testing.T, m, n int) *addrBook {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	// add m old addresses
+	randAddrs := randNetAddressPairs(t, m)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+		book.MarkGood(addr.addr)
+	}
+	// add n new addresses
+	randAddrs = randNetAddressPairs(t, n)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	return book
+}
+
+// Computes n * bias/100
+func numExpectedAddresses(bias, n int) int {
+	if n == 0 {
+		return 0
+	}
+	return int(math.Round((float64(bias) / float64(100)) * float64(n)))
+}
+
+// Counts the number of Old and New addresses
+func countOldAndNewInAddrs(addrs []*p2p.NetAddress, book *addrBook) (nold, nnew int) {
+	nold = 0
+	nnew = 0
+
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		if book.IsGood(addr) {
```

Comment:
good != old

Reply:
How so?

```
func (a *addrBook) IsGood(addr *p2p.NetAddress) bool {
	a.mtx.Lock()
	defer a.mtx.Unlock()

	return a.addrLookup[addr.ID].isOld()
}

```

Reply:
What I am trying to say is it might change in the future (e.g. good addr - old addr whos been old for a month), so maybe it's better to create a `IsOld(addr *p2p.NetAddress)` function.

Reply:
Not sure what to do here. The IsGood() function is only used by addrbook_test.go at this point and not only in the newly added tests. I think it's a good opportunity to change its name to isOld() as I found it confusing. Let me know.

Reply:
I think we can leave it as IsGood because we use Good/Bad on the public interface and old/new internally. It's true that eventually they may diverge and there could be different levels of Good, but I think we can leave for now.